frontend/ui/presets.lua

```diff
@@ -0,0 +1,357 @@
+It handles creation, loading, updating, and deletion of presets, as well as menu generation.
+
+    Usage example:
+        local Presets = require("ui/presets")
+
+        -- In your module:
+        function MyModule:buildPreset()
+            return {
+                -- Return a table with the settings you want to save
+                setting1 = self.setting1,
+                setting2 = self.setting2,
+            }
+        end
+
+        function MyModule:loadPreset(preset)
+            -- Apply the preset settings to your module
+            self.setting1 = preset.setting1
+            self.setting2 = preset.setting2
+            -- Update UI or other necessary changes
+            self:refresh()
+        end
+
+        -- To create menu items for presets:
+        function MyModule:genPresetMenuItemTable()
+            return Presets:genModulePresetMenuTable(
+                self,
+                "my_module_presets", -- preset key for settings
+                _("Create new preset") -- optional custom text
+            )
+        end
+
+        -- To create a new preset:
+        function MyModule:createPresetFromCurrentSettings(touchmenu_instance)
+            return Presets:createModulePreset(self, touchmenu_instance, "my_module_presets")
+        end
+
+        -- To load a preset by name:
+        function MyModule:onLoadPreset(preset_name)
+            return Presets:onLoadPreset(self, preset_name, "my_module_presets", true)
+        end
+
+        -- To get list of available presets:
+        function MyModule.getPresets()
+            return Presets:getPresets("my_module_presets")
+        end
+
+Required module methods:
+- buildPreset(): returns a table with the settings to save
+- loadPreset(preset): applies the settings from the preset table
+
+The preset system handles:
+- Saving/loading presets to/from G_reader_settings
+- Creating and managing preset menu items
+- User interface for creating/updating/deleting presets
+- Notifications when presets are loaded/updated
+- Adding actions to Dispatcher for loading presets through gestures/hotkeys/profiles
+--]]
+
+local ConfirmBox = require("ui/widget/confirmbox")
+local Event = require("ui/event")
+local InfoMessage = require("ui/widget/infomessage")
+local InputDialog = require("ui/widget/inputdialog")
+local Notification = require("ui/widget/notification")
+local UIManager = require("ui/uimanager")
+local ffiUtil = require("ffi/util")
+local T = require("ffi/util").template
+local _ = require("gettext")
+
+local Presets = {}
+
+function Presets:createPresetFromCurrentSettings(touchmenu_instance, preset_name_key, buildPresetFunc, genPresetMenuItemTableFunc)
+    self:editPresetName({},
+        function(entered_preset_name, dialog_instance) -- this is the on_confirm_callback.
+        -- It uses preset_name_key, buildPresetFunc, and genPresetMenuItemTableFunc from the outer scope.
+        local presets = G_reader_settings:readSetting(preset_name_key, {})
```

Comment:
As a UI abstraction, the Presets module shouldn't know anything about the presets storage.
The Presets module should get a "presets" table and work with it.
Moreover, not a separate "presets" table, but a module that contains the "presets" table, necessary properties (like `action_name`), and methods to work with presets (like buildPreset etc.).